src/main/java/com/marklogic/contentpump/ThreadManager.java

```diff
@@ -0,0 +1,372 @@
+ * @author vzhang
+ */
+public class ThreadManager implements ConfigConstants {
+    public static final Log LOG = LogFactory.getLog(ThreadManager.class);
+    public static final String SERVER_MAX_THREADS_QUERY =
+        "import module namespace hadoop = " +
+        "\"http://marklogic.com/xdmp/hadoop\" at \"/MarkLogic/hadoop.xqy\";\n" +
+        "let $f := " +
+        "fn:function-lookup(xs:QName('hadoop:get-port-max-threads'),0)\n" +
+        "return if (exists($f)) then $f() else 0";
+    private LocalJob job;
+    private Command cmd;
+    private Configuration conf;
+    private final ScheduledExecutorService scheduler;
+    private ThreadPoolExecutor pool;
+    private List<LocalMapTask> taskList = new ArrayList<>();
+    private List<Future<Object>> taskFutureList = new ArrayList<>();
+    // minimally required thread per task defined by the job
+    private int minThreads = 1;
+    private int curServerThreads;
+
+    /**
+     * Static variables set by checkOutputFormat
+     */
+    private static int newServerThreads;
+    // Whether mlcp is running against a load balancer
+    private static boolean restrictHosts = false;
+
+    /**
+     * Command line options
+     */
+    private int threadCount;
+    private int threadsPerSplit;
+    private static double maxThreadPercentage = 1;
+
+    public ThreadManager(LocalJob job, CommandLine cmdline, Command cmd) {
+        this.job = job;
+        this.cmd = cmd;
+        this.conf = job.getConfiguration();
+        this.minThreads = conf.getInt(CONF_MIN_THREADS, minThreads);
+        this.scheduler = Executors.newScheduledThreadPool(1);
+
+        // Parse command line options
+        if (cmdline.hasOption(THREAD_COUNT)) {
+            threadCount = Integer.parseInt(
+                cmdline.getOptionValue(THREAD_COUNT));
+        }
+        if (cmdline.hasOption(THREADS_PER_SPLIT)) {
+            threadsPerSplit = Integer.parseInt(
+                cmdline.getOptionValue(THREADS_PER_SPLIT));
+        }
+        if (cmdline.hasOption(MAX_THREAD_PERCENTAGE)) {
+            maxThreadPercentage = Double.parseDouble(
+                cmdline.getOptionValue(MAX_THREAD_PERCENTAGE));
+        }
+    }
+
+    /**
+     * Query the server stack to get the maximum available thread count.
+     * @param cs
+     * @throws IOException
+     */
+    public static void queryServerMaxThreads(ContentSource cs)
+        throws IOException {
+        Session session = null;
+        ResultSequence result = null;
+        try {
+            session = cs.newSession();
+            AdhocQuery query = session.newAdhocQuery(SERVER_MAX_THREADS_QUERY);
+            RequestOptions options = new RequestOptions();
+            options.setDefaultXQueryVersion("1.0-ml");
+            query.setOptions(options);
+            result = session.submitRequest(query);
+
+            if (result.hasNext()) {
+                ResultItem item = result.next();
+                newServerThreads = (int)(maxThreadPercentage *
```

Comment:
There's no maxthread cap anymore, the client might not be able to keep up with large clusters? mlcp users should be able to cap newServerThreads.